Plugins/Rename/Rename.cpp

```diff
@@ -95,74 +110,107 @@ CNWSPlayer *Rename::player(Types::ObjectID playerId)
         CNWSCreature* pCreature = server->GetCreatureByGameObjectID(pcObjectID);
         
         if (pCreature && !lsFirstName.empty() && ( bOriginal || !pCreature->m_sDisplayName.IsEmpty()))
         {
+            if (pPlayer != nullptr || static_cast<bool>(*g_plugin->GetServices()->m_perObjectStorage->Get<int>(pcObjectID, playerNameOverrideStateKey)) //if the pPlayer pointer is null (from a party invite) or the player name override is default...
+                                   || !static_cast<bool>((server->GetCreatureByGameObjectID(pPlayer->m_oidNWSObject))->m_pStats->m_bIsDM) //or if pPlayer is a DM...
+                                   || pPlayer->m_oidNWSObject != pcObjectID) //or pCreature is pPlayer's creature object then skip changing the player name, since it's not needed or desired.
+            {
+                if (bOriginal)
+                {
+                    Globals::AppManager()->m_pServerExoApp->GetNetLayer()->GetPlayerInfo(client->m_nPlayerID)->m_sPlayerName = CExoString((*g_plugin->GetServices()->m_perObjectStorage->Get<std::string>(pcObjectID, playerNameKey)).c_str());
+                }
+                else
+                {
+                    switch((*g_plugin->GetServices()->m_perObjectStorage->Get<int>(pcObjectID, playerNameOverrideStateKey)))
+                    {
+                        case NWNX_RENAME_PLAYERNAME_OBFUSCATE  : Globals::AppManager()->m_pServerExoApp->GetNetLayer()->GetPlayerInfo(client->m_nPlayerID)->m_sPlayerName = CExoString(GenerateRandomPlayerName(7).c_str());
+                                  break;
+                        case NWNX_RENAME_PLAYERNAME_OVERRIDE  : Globals::AppManager()->m_pServerExoApp->GetNetLayer()->GetPlayerInfo(client->m_nPlayerID)->m_sPlayerName = CExoString(lsFirstName.c_str());
+                                  break;
+                    }
+                }
+            }
             pCreature->m_pStats->m_lsFirstName = ContainString(lsFirstName);
             pCreature->m_pStats->m_lsLastName = ContainString(lsLastName);
         }
     }
 }
 
-std::string Rename::ExtractString(CExoLocString locStr)
+std::string Rename::ExtractString(CExoLocString& locStr)
 {
     CExoString str;
     locStr.GetStringLoc(0,&str,0);
     return std::string(str.CStr());
 }
 
-CExoLocString Rename::ContainString(std::string str)
+CExoLocString Rename::ContainString(const std::string& str)
 {
     CExoLocString locStr;
     locStr.AddString(0,CExoString(str.c_str()),0);
     return locStr;
 }
 
+std::string Rename::GenerateRandomPlayerName(size_t length)
+{
+    static std::mt19937 rngEngine(std::random_device{}());
+    static const std::string charSet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"; 
+    std::uniform_int_distribution<int> distribution(0, charSet.length() - 1);
+    std::string randomPlayername;
+    for (size_t i = 0; i < length; ++i)
+    {
+        randomPlayername += charSet[distribution(rngEngine)];
```

Comment:
Each time you do this `+=`, new storage for the string is created, one byte larger than the previous, and previous is destroyed. So you do `length` allocations/deletes. Add `randomPlayername.reserve(length+1)` before the loop to reserve all the storage immediately.